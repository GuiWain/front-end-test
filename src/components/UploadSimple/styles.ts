import styled, { css, DefaultTheme } from 'styled-components'

import { motion, AnimatePresence } from 'framer-motion'
import UseImage from 'next/image'

export type WrapperProps = {
  error?: boolean
  disabled?: boolean
}

export const Animate = styled(AnimatePresence)``
export const Motion = styled(motion.div)``

export const DownloadIcon = styled(UseImage).attrs({
  src: '/img/download.svg',
  width: '16px',
  height: '16px',
  alt: 'Download Icon'
})`
  cursor: pointer;
`
export const CloseIcon = styled(UseImage).attrs({
  'data-testid': 'CloseIcon',
  src: '/img/close.svg',
  width: '12px',
  height: '12px',
  alt: 'Close Icon'
})`
  cursor: pointer;
  margin-left: auto;
`

const wrapperModifiers = {
  error: (theme: DefaultTheme) => css`
    ${InputFile}, ${DivInput}, ${ButtonLabel} {
      border-color: ${theme.colors.errorMedium};

      &:focus-within {
        border: 0.1rem solid ${theme.colors.errorMedium};
      }
    }

    ${LabelText} {
      color: ${theme.colors.errorMedium};
    }
  `,
  disabled: (theme: DefaultTheme) => css`
    ${InputFile}, ${DivInput}, ${ButtonLabel}, ${LabelText}, ${FileName} {
      cursor: not-allowed;
      color: ${theme.colors.secondaryMedium};
    }
  `
}

export const Wrapper = styled.div<WrapperProps>`
  ${({ theme, disabled, error }) => css`
    background: transparent;

    ${error && wrapperModifiers.error(theme)}
    ${disabled && wrapperModifiers.disabled(theme)}
  `}
`

export const DivInput = styled.div`
  ${({ theme }) => css`
    background-color: ${theme.colors.white};
    border: 0.1rem solid ${theme.colors.secondaryLight};
    border-radius: ${theme.border.smallRadius};

    display: flex;
    justify-content: flex-start;
    align-items: center;
    gap: 0.6rem;
  `}
`

export const ButtonLabel = styled.label`
  ${({ theme }) => css`
    background-color: ${theme.colors.neutralMedium};
    border-radius: 0.3rem;
    padding: 0.5rem 1rem;

    font: normal normal ${theme.font.semiBold} ${theme.font.sizes.xxsmall} /
      2.8rem ${theme.font.family};
    color: ${theme.colors.secondaryDark};
    letter-spacing: 0rem;
    text-align: left;
    cursor: pointer;
  `}
`

export const FileName = styled.span`
  ${({ theme }) => css`
    font: normal normal ${theme.font.regular} ${theme.font.sizes.xxsmall} /
      2.8rem ${theme.font.family};
    color: ${theme.colors.neutralDark};
    letter-spacing: 0rem;
    text-align: left;
    cursor: pointer;
  `}
`

export const InputFile = styled.input.attrs({
  type: 'file',
  accept:
    'application/pdf, application/msword, application/vnd.ms-excel, image/*'
})`
  display: none;
`

export const FileList = styled.div`
  width: 100%;
  margin: 0 auto;
`

export const Item = styled.div`
  ${({ theme }) => css`
    display: flex;
    justify-items: center;
    align-items: center;
    gap: 0.7rem;
    margin: 0.5rem 0;
    padding: 1rem 2rem;

    text-align: left;
    font: normal normal ${theme.font.regular} ${theme.font.sizes.xxsmall} / 2rem
      ${theme.font.family};
    letter-spacing: -0.035rem;
    color: ${theme.colors.infoDark};

    background: ${theme.colors.white} 0% 0% no-repeat padding-box;
    border: 0.1rem solid ${theme.colors.secondaryLight};
    border-radius: ${theme.border.smallRadius};
  `}
`

export const LabelText = styled.p`
  ${({ theme }) => css`
    font: normal normal ${theme.font.semiBold} ${theme.font.sizes.xsmall} / 4rem
      ${theme.font.family};

    letter-spacing: -0.006rem;
    text-align: left;
    color: ${theme.colors.secondaryDark};

    margin: 0;
  `}
`

export const Error = styled.p`
  ${({ theme }) => css`
    color: ${theme.colors.errorMedium};
    font-size: ${theme.font.sizes.xxxsmall};
    margin: 0.5rem 0;
  `}
`
