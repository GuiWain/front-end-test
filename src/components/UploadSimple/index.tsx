import { useState, useEffect } from 'react'
import * as S from './styles'

import { v4 as uuidv4 } from 'uuid'

export type FileListValues = {
  id: string
  files: File
}

export type UploadSimpleProps = {
  getValues?: (field: FileListValues[]) => void
  label?: string
  button?: string
  name?: string
  multiple?: boolean
  disabled?: boolean
  error?: string
}

const UploadSimple = ({
  getValues,
  button,
  label,
  multiple = true,
  name,
  disabled = false,
  error
}: UploadSimpleProps) => {
  const [fileList, setFileList] = useState<FileListValues[]>([])

  useEffect(() => {
    if (getValues) {
      getValues(fileList)
    }
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [fileList])

  const onChange = (e: React.ChangeEvent<HTMLInputElement>) => {
    const files = e.target.files!
    const arrayFiles = Object.values(files)
    const array = arrayFiles.map((item) => ({ files: item, id: uuidv4() }))

    setFileList((s) => [...s, ...array])
  }

  const removeFile = (index: number) => {
    const newFiles = [...fileList]

    newFiles.splice(index, 1)
    setFileList(newFiles)
  }

  return (
    <S.Wrapper error={!!error} disabled={disabled}>
      {!!label && <S.LabelText>{label}</S.LabelText>}
      <S.DivInput>
        {!!button && <S.ButtonLabel htmlFor={name}>{button}</S.ButtonLabel>}
        <S.FileName>
          {fileList?.length && fileList?.length === 1
            ? fileList.length + ' Selected file'
            : null}
          {fileList?.length && fileList?.length > 1
            ? fileList.length + ' Selected files'
            : null}
          {fileList?.length ? null : 'No file selected'}
        </S.FileName>
        <S.InputFile
          id={name}
          name={name}
          onChange={(e) => onChange(e)}
          multiple={multiple}
          disabled={disabled}
        />
      </S.DivInput>
      <S.Animate>
        {!!fileList.length && (
          <S.FileList>
            <S.Animate>
              {fileList.map((item, index) => (
                <S.Motion
                  key={item.id}
                  transition={{ ease: 'easeInOut', duration: 0.5 }}
                  initial={{ opacity: 0, y: 10 }}
                  animate={{ opacity: 1, y: 0 }}
                  exit={{ opacity: 0, y: -10 }}
                >
                  <S.Item>
                    <S.DownloadIcon />
                    {item.files.name}
                    <S.CloseIcon onClick={() => removeFile(index)} />
                  </S.Item>
                </S.Motion>
              ))}
            </S.Animate>
          </S.FileList>
        )}
      </S.Animate>
      {!!error && <S.Error>{error}</S.Error>}
    </S.Wrapper>
  )
}

export default UploadSimple
