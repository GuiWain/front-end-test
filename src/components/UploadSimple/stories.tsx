import { Story, Meta } from '@storybook/react'
import UploadSimple, { UploadSimpleProps } from '.'

export default {
  title: 'Inputs/UploadSimple',
  component: UploadSimple,
  argTypes: {
    label: {
      type: 'string'
    },
    button: {
      type: 'string'
    },
    name: {
      type: 'string'
    },
    disabled: {
      control: {
        type: 'radio',
        options: [true, false]
      }
    },
    multiple: {
      control: {
        type: 'radio',
        options: [true, false]
      }
    },
    error: {
      type: 'string'
    }
  }
} as Meta

export const Default: Story<UploadSimpleProps> = (args) => (
  <div style={{ width: '40%', margin: '2rem auto' }}>
    <UploadSimple {...args} />
  </div>
)

Default.args = {
  label: 'Select files to upload',
  button: 'Choose file',
  name: 'inputFile',
  multiple: true,
  disabled: false
}

export const WithError: Story<UploadSimpleProps> = (args) => (
  <div style={{ width: '40%', margin: '2rem auto' }}>
    <UploadSimple {...args} />
  </div>
)

WithError.args = {
  label: 'Select files to upload',
  button: 'Choose file',
  multiple: true,
  name: 'inputFile',
  error: 'Required field',
  disabled: false
}
