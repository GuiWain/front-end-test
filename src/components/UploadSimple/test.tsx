import userEvent from '@testing-library/user-event'
import { render, waitFor, screen, act, fireEvent } from 'utils/test-utils'

import UploadSimple from '.'

describe('<UploadSimple />', () => {
  it('should render the UploadSimple', () => {
    const { container } = render(<UploadSimple />)

    expect(container.firstChild).toMatchSnapshot()
  })

  it('must upload one file', () => {
    render(<UploadSimple button="Choose file" name="inputFile" />)

    const file = [new File(['foo'], 'foo.pdf', { type: 'application/pdf' })]

    const input = screen.getByLabelText('Choose file')
    fireEvent.change(input, {
      target: { files: file }
    })
  })

  it('must upload two file', async () => {
    render(<UploadSimple button="Choose file" name="inputFile" />)

    await waitFor(() => {
      const files = [
        new File(['foo'], 'foo.pdf', { type: 'application/pdf' }),
        new File(['bar'], 'bar.png', { type: 'image/png' })
      ]

      const input = screen.getByLabelText('Choose file')
      userEvent.upload(input, files)
    })

    expect(screen.getByText('foo.pdf')).toBeInTheDocument()
    expect(screen.getByText('bar.png')).toBeInTheDocument()
  })

  it('must remove a file from the list', async () => {
    render(<UploadSimple button="Choose file" name="inputFile" />)

    await waitFor(() => {
      const files = [new File(['foo'], 'foo.pdf', { type: 'application/pdf' })]

      const input = screen.getByLabelText('Choose file')
      userEvent.upload(input, files)
    })

    act(() => {
      const closeIcon = screen.getAllByTestId('CloseIcon')
      userEvent.click(closeIcon[0])
      expect.anything()
    })
  })

  it('should render the UploadSimple with label', () => {
    render(<UploadSimple label="Choose file" />)

    const label = screen.getByText('Choose file')
    expect(label).toBeInTheDocument()
  })

  it('should render the UploadSimple disabled', () => {
    render(
      <UploadSimple disabled={true} button="Choose file" name="inputFile" />
    )

    const input = screen.getByLabelText('Choose file')
    expect(input).toBeDisabled()
  })

  it('should render the UploadSimple with error', () => {
    render(
      <UploadSimple
        error="Required field"
        button="Choose file"
        name="inputFile"
      />
    )

    const error = screen.getByText('Required field')
    expect(error).toBeInTheDocument()
  })

  it('should render the UploadSimple with getValues', async () => {
    const getValues = jest.fn()
    render(
      <UploadSimple
        button="Choose file"
        name="inputFile"
        getValues={getValues}
      />
    )

    await waitFor(() => {
      const file = [new File(['foo'], 'foo.pdf', { type: 'application/pdf' })]

      const input = screen.getByLabelText('Choose file')
      userEvent.upload(input, file)
    })

    expect(getValues).toHaveBeenCalled()
  })
})
