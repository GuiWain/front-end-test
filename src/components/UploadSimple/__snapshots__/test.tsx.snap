// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<UploadSimple /> should render the UploadSimple 1`] = `
.c0 {
  background: transparent;
}

.c1 {
  background-color: #FFFFFF;
  border: 0.1rem solid #CCC8C8;
  border-radius: 0.4rem;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: start;
  -webkit-justify-content: flex-start;
  -ms-flex-pack: start;
  justify-content: flex-start;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  gap: 0.6rem;
}

.c2 {
  font: normal normal 400 1.4rem / 2.8rem 'Open Sans','Roboto','Helvetica Neue',sans-serif;
  color: #9D9D9D;
  -webkit-letter-spacing: 0rem;
  -moz-letter-spacing: 0rem;
  -ms-letter-spacing: 0rem;
  letter-spacing: 0rem;
  text-align: left;
  cursor: pointer;
}

.c3 {
  display: none;
}

<div
  class="c0"
>
  <div
    class="c1"
  >
    <span
      class="c2"
    >
      No file selected
    </span>
    <input
      accept="application/pdf, application/msword, application/vnd.ms-excel, image/*"
      class="c3"
      multiple=""
      type="file"
    />
  </div>
</div>
`;
